lib/stdlib/src/uri_string.erl

```diff
@@ -375,6 +375,98 @@ recompose(Map) ->
     end.
 
 
+%%-------------------------------------------------------------------------
+%% Resolve URIs
+%%-------------------------------------------------------------------------
+-spec resolve(RefURI, BaseURI) -> TargetURI when
+      RefURI :: uri_string() | uri_map(),
+      BaseURI :: uri_string() | uri_map(),
+      TargetURI :: uri_string()
+                 | error().
+resolve(URIMap, BaseURIMap) ->
+    resolve(URIMap, BaseURIMap, []).
+
+
+-spec resolve(RefURI, BaseURI, Options) -> TargetURI when
+      RefURI :: uri_string() | uri_map(),
+      BaseURI :: uri_string() | uri_map(),
+      Options :: [return_map],
+      TargetURI :: uri_string() | uri_map()
+                 | error().
+resolve(URIMap, BaseURIMap, Options) when is_map(URIMap) ->
+    case resolve_map(URIMap, BaseURIMap) of
+        TargetURIMap when is_map(TargetURIMap) ->
+            case Options of
+                [return_map] ->
+                    TargetURIMap;
+                [] ->
+                    recompose(TargetURIMap)
+            end;
+        Error ->
+            Error
+    end;
+resolve(URIString, BaseURIMap, Options) ->
+    case parse(URIString) of
+        URIMap when is_map(URIMap) ->
+            resolve(URIMap, BaseURIMap, Options);
+        Error ->
+            Error
+    end.
+
+resolve_map(URIMap=#{scheme := _}, _) ->
+    normalize_path_segment(URIMap);
+resolve_map(URIMap, BaseURIMap) when is_map(BaseURIMap) ->
+    resolve_map(URIMap, BaseURIMap, resolve_path_type(URIMap));
+resolve_map(URIMap, BaseURIString) ->
+    case parse(BaseURIString) of
+        BaseURIMap when is_map(URIMap) ->
+            resolve_map(URIMap, BaseURIMap, resolve_path_type(URIMap));
+        Error ->
+            Error
+    end.
+
+resolve_path_type(URIMap) ->
+    case iolist_to_binary(maps:get(path, URIMap, <<>>)) of
+        <<>> -> empty_path;
+        <<$/,_/bits>> -> absolute_path;
+        _ -> relative_path
+    end.
+
+resolve_map(URI=#{host := _}, #{scheme := Scheme}, _) ->
+	normalize_path_segment(URI#{scheme => Scheme});
+resolve_map(URI, BaseURI, empty_path) ->
+	Keys = case maps:is_key(query, URI) of
+		true -> [scheme, userinfo, host, port, path];
+		false -> [scheme, userinfo, host, port, path, query]
+	end,
+	maps:merge(URI, maps:with(Keys, BaseURI));
+resolve_map(URI, BaseURI, absolute_path) ->
+	normalize_path_segment(maps:merge(
+        URI,
+        maps:with([scheme, userinfo, host, port], BaseURI)));
+resolve_map(URI=#{path := Path}, BaseURI, relative_path) ->
+	normalize_path_segment(maps:merge(
+		URI#{path => merge_paths(Path, BaseURI)},
+		maps:with([scheme, userinfo, host, port], BaseURI))).
+
+merge_paths(Path, #{path := BasePath0}) ->
+    case iolist_size(BasePath0) of
+        0 ->
+            merge_paths_absolute(Path);
```

Comment:
```
If the base URI has a defined authority component and an empty
path, then return a string consisting of "/" concatenated with the
reference's path; otherwise,
```
```
1> uri_string:resolve("a/b/c","").
"/a/b/c"
```
I've found this special case with an empty BaseURI where I think the correct result should be "a/b/c".

Reply:
There's a mistake but your interpretation is wrong. It says if the base URI *has* an authority component, so this does not include `""`. Note that based on the algorithm the `merge_path` function does not seem to ever be allowed to return a relative path.

The mistake is that when refactoring I dropped the check on the authority component entirely. I will add it back.

Reply:
You are right, it cannot return _just_ a relative path, but we need to add more checks even for the base URI. According to chapter 5.2.1, the scheme component is required to be present in a base URI.
```
5.2.1.  Pre-parse the Base URI

   The base URI (Base) is established according to the procedure of
   Section 5.1 and parsed into the five main components described in
   Section 3.  Note that only the scheme component is required to be
   present in a base URI; the other components may be empty or
   undefined.  A component is undefined if its associated delimiter does
   not appear in the URI reference; the path component is never
   undefined, though it may be empty.
```
This requirement should be fulfilled with those in chapter 5.2.3:

```
5.2.3.  Merge Paths

   The pseudocode above refers to a "merge" routine for merging a
   relative-path reference with the path of the base URI.  This is
   accomplished as follows:

   o  If the base URI has a defined authority component and an empty
      path, then return a string consisting of "/" concatenated with the
      reference's path; otherwise,

   o  return a string consisting of the reference's path component
      appended to all but the last segment of the base URI's path (i.e.,
      excluding any characters after the right-most "/" in the base URI
      path, or excluding the entire base URI path if it does not contain
      any "/" characters).
```

So my previous example should return an error tuple instead as the base URI is not valid. Then we have the case when the base URI consists of only the scheme component e.g. ```http:```. In that case we cannot apply the first paragraph of 5.2.3 as the base URI has an empty path but the authority component is missing. We can apply the second paragraph and exclude the entire base URI path as it does not contain any "/" characters. So we get a target URI with a scheme and a relative path:

```
1> uri_string:resolve("a/b/c","http:").
"http:a/b/c"
```
I think this could be added to the abnormal examples.

Reply:
We need tests with many different types of base URIs. It's weird that they didn't add any.

By the way what does `"http:a/b/c"` even mean? For some URI schemes I suppose they would be valid, but for `http` it's invalid according to https://tools.ietf.org/html/rfc7230#section-2.7.1

```
   A sender MUST NOT generate an "http" URI with an empty host
   identifier.  A recipient that processes such a URI reference MUST
   reject it as invalid.
```

Do we want to special case `http`, `https` and perhaps other schemes and return an error for them if the base URI has no `host` component?

Reply:
I'm more inclined to have special handling for at least `http `and `https`. RFC3986 includes for instance the scheme based normalization for `http`, `https`, `ftp`, `ssh`, `sftp` and `tftp`. We also took a pragmatic approach for the handling of urlencoded query strings based on HTML 5.0 and HTML 5.2 (WHAT WG URL). I think having a special check for `http `and `https ` would have more practical benefits compared to keeping the module's knowledge of the `http` scheme on a bare minimum.

Reply:
Hmm but then there's the `http:g` example where `http:g` must be returned, so as far as HTTP is concerned perhaps a separate function would be better to check that you have a URI that's valid according to the scheme. I don't know. I'm fine checking it in user code for now and perhaps we can discuss this again after some real world experience.

I will add tests with different types of base URIs for now.